mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala

```diff
@@ -474,6 +487,75 @@ class LinearRegressionSummary private[regression] (
     predictions.select(t(col(predictionCol), col(labelCol)).as("residuals"))
   }
 
+  /** Number of instances in DataFrame predictions */
+  lazy val numInstances: Long = predictions.count()
+
+  /** Degrees of freedom */
+  private val degreesOfFreedom: Long = if (model.getFitIntercept) {
+    numInstances - model.coefficients.size - 1
+  } else {
+    numInstances - model.coefficients.size
+  }
+
+  /**
+   * The weighted residuals, the usual residuals rescaled by
+   * the square root of the instance weights.
+   */
+  lazy val devianceResiduals: Array[Double] = {
```

Comment:
@jkbradley There is "residuals" already exist, so I call it ```devianceResiduals```. I agree your opinion about adding other types of residuals later, so I think we can try to combine the two functions into one with different arguments. We also need do some code clean up for ```LinearRegressionSummary``` due to redundant arguments, I can finish it in a follow up PR. @mengxr